addons/test_portal/tests/test_mail_links.py

```diff
@@ -0,0 +1,137 @@
+from collections import defaultdict
+from itertools import product
+
+from lxml import html
+from werkzeug.test import Client
+from werkzeug.urls import url_encode, url_parse
+from werkzeug.wrappers import BaseResponse
+
+from odoo import http
+from odoo.tests import HttpCase
+from odoo.tools import mute_logger
+
+
+class TestMailAccess(HttpCase):
+    def _cleanup(self, location):
+        # we want the absolute path but not the absolute URL, strip out bullshit
+        # scheme & domain, as well as fragment as Werkzeug's client doesn't
+        # support those (TBF they're a client-only concern)
+        return url_parse(location).replace(scheme='', netloc='', fragment='').to_url()
+
+    def do_login(self, client, body, user):
+        h = html.fromstring(body)
+        f = h.find('.//form[@class="oe_login_form"]')
+        fields = {
+            field.get('name'): field.get('value')
+            for field in f.xpath('.//input[@name] | .//button[@name]')
+            if field.get('value')
+        }
+        fields['login'] = fields['password'] = user
+        r = client.post(f.get('action'), data=fields, environ_base={
+            'REMOTE_HOST': 'localhost',
+            'REMOTE_ADDR': '127.0.0.1',
+        })
+        return r.headers['Location'], self._cleanup(r.headers['Location'])
+
+    def do_flow(self, model, user, *, login, token):
+        params = {}
+        Model = self.env[model]
+        if token and 'access_token' in Model._fields:
+            params = {'access_token': '12345'}
+
+        obj = Model.create(params)
+        c = Client(http.root, BaseResponse)
+        if login:
+            self.authenticate(user, user)
+            c.set_cookie('localhost', key='session_id', value=self.session.sid, httponly=True)
+        else:
+            self.env.cr.flush()
+            self.env.cr.clear()
+        location = url = '/mail/view?' + url_encode({
+            'model': model,
+            'res_id': obj.id,
+            **params
+        })
+        while True:
+            r = c.get(url)
+            self.assertLess(r.status_code, 400, f'{url} = {r.status}')
```

Comment:
Isn't it equivalent to result.ok ?

Reply:
And of course after actually trying it it doesn't work because `r` is a `werkzeug.wrappers.Response`, not a `requests.Response`, and so does not *have* an `ok` property.